docs/guides/contexts_expressions.md

```diff
@@ -0,0 +1,390 @@
+```
+
+A string expression can be converted to a napari expression with
+`parse_expression`:
+
+```python
+In [6]: from napari.utils.context import parse_expression
+
+In [7]: expr = parse_expression("x > 5 and y == 'hello'")
+
+In [8]: expr
+Out[8]:
+BoolOp(
+  op=And(),
+  values=[
+    Compare(
+      left=Name(id='x', ctx=Load()),
+      ops=[Gt()],
+      comparators=[Constant(value=5)]
+    ),
+    Compare(
+      left=Name(id='y', ctx=Load()),
+      ops=[Eq()],
+      comparators=[Constant(value='hello')]
+    )
+  ]
+)
+```
+
+The expression object can be evaluated using `eval`
+
+```python        
+In [9]: expr.eval({'x': 7, 'y': 'hello'})
+Out[9]: True
+```
+
+and it can also be combined with other expressions & constants using
+**operators**:
+
+```python
+In [10]: new_expr = expr & (7 > 10)  # always False
+
+In [11]: new_expr.eval({'x': 7, 'y': 'hello'})
+Out[12]: False
+```
+
+### napari context keys
+
+To capture napari-specific conditions, napari will declare special string
+**names** that can be used in a napari expression.  Taking the example above, a
+plugin might only want to provide a function if “the active layer has at least 3
+dimensions”. For this, napari recognizes the variable `active_layer_ndim` used
+in an expression. In a plugin manifest, the plugin can provide a *when clause*
+to enable/disable a given command:
+
+```yaml
+command:
+   id: myplugin.my_command
+   when: active_layer_ndim >= 3
+```
+
+Some example context key names (currently) include:
+
+| Name     | Description |
+| -------- | -------- |
+| `layers_selection_count`  | Number of layers currently selected  | 
+| `all_layers_linked`  | True when all selected layers are linked  | 
+| `active_layer_is_rgb`  | True when the active layer is RGB  | 
+| `active_layer_type`  | Lowercase name of active layer type, or None if no layer is active.  | 
+| `only_images_selected`  | True when there is at least one selected layer and all selected layers are images  | 
+| `active_layer_ndim` | Number of dimensions in the active layer, or `None` if nothing is active |
+...  many more
+
+
+### `ContextKey` objects
+
+To track the special **name** strings that can be used in expressions, napari
+has the `ContextKey` class.  Instances of `ContextKey` subclass from
+[`ast.Name`](https://docs.python.org/3/library/ast.html#ast.Name) and represent
+a variable name in an expression. Additionally, they have a `description` that
+can be used in documentation, a `default_value`, and a `getter` function that
+can be called to retrieve the current value (the parameters passed to the
+`getter` will depend on the context key... see [Updating
+Contexts](#updating-contexts) below for more).
+
+In order to keep related keys together, `ContextKey` instances will usually be
+declared as class attributes on a `ContextNamespace` class:
+
+```python
+# all of the getters here receive an instance of viewer.layers.selection
+
+class LayerListContextKeys(ContextNamespace):
```

Comment:
Hum, thinking about it I don't like the `LayerListContextKeys` name for the class. It mixes the Terminologies of Keys and Namespace. Would it be better to name it `LayerListContextNamespace` ?

Reply:
let's address that elsewhere.  This isn't something that anyone outside of napari will use.  easy to change later